src/chainstate/stacks/index/file.rs

```diff
@@ -139,35 +191,31 @@ impl TrieFile {
     /// Get a copy of the path to this TrieFile.
     /// If in RAM, then the path will be ":memory:"
     pub fn get_path(&self) -> String {
-        match self {
-            TrieFile::RAM(_) => ":memory:".to_string(),
-            TrieFile::Disk(ref disk) => disk.path.clone(),
-        }
-    }
-
-    /// Instantiate a TrieFile, given the associated DB path.
-    /// If path is ':memory:', then it'll be an in-RAM TrieFile.
-    /// Otherwise, it'll be stored as `$db_path.blobs`.
-    pub fn from_db_path(path: &str, readonly: bool) -> Result<TrieFile, Error> {
-        if path == ":memory:" {
-            Ok(TrieFile::new_ram(readonly))
-        } else {
-            let blob_path = format!("{}.blobs", path);
-            TrieFile::new_disk(&blob_path, readonly)
-        }
+        self.path.clone()
     }
 
     /// Append a new trie blob to external storage, and add the offset and length to the trie DB.
     /// Return the trie ID
     pub fn store_trie_blob<T: MarfTrieId>(
         &mut self,
+        compression_type: &BlobCompressionType,
         db: &Connection,
         bhh: &T,
         buffer: &[u8],
     ) -> Result<u32, Error> {
-        let offset = self.append_trie_blob(db, buffer)?;
-        test_debug!("Stored trie blob {} to offset {}", bhh, offset);
-        trie_sql::write_external_trie_blob(db, bhh, offset, buffer.len() as u64)
+        let result = self.append_trie_blob(compression_type, db, buffer, None)?;
+        //eprintln!("Stored trie blob {} to offset {} with length {:?}", bhh, result.offset, result.storage_size);
+        let block_id = trie_sql::write_external_trie_blob(
+            db,
+            bhh,
+            result.offset,
+            result.storage_size as u64,
+            compression_type,
+        )?;
+
+        self.decompressed_lru.put(block_id, buffer.to_vec());
```

Comment:
It seems like `self.decompressed_lru` is going to get pretty big.  Each serialized trie is between 10**5 and 10**7 bytes.

Also, it seems like the LRU cache stores up to 256 tries, right?  If so, then while this would save disk I/O accesses and decompression work for walking trie nodes in recently-accessed tries, I worry about cache thrashing.  For example, someone who wanted to DoS the network could craft a malicious transaction that deliberately loaded a large number of infrequently-accessed tries during its execution, causing mining to slow down and causing validation to slow down (possibly by a considerable margin).

Reply:
Yeah it's hardcoded at 256 right now but have a point above to make it configurable (cache size of zero would effectively disable the cache), and the default could be a much more conservative number.

Since this is an opt-in feature for node operators it would be impossible for a potential attacker to know the configurations of each node to plan such an attack (is compression even enabled? If so, which algorithm? Is their LRU cache enabled and if so how large is their cache? Spec of the node, etc.).

But I agree, potential attack vectors should be analyzed :)